ctapipe/io/zfits.py

```diff
@@ -4,109 +4,96 @@
 
 This requires the protozfitsreader python library to be installed
 """
-import logging
+from numpy import ndarray
 
-import numpy as np
-import numpy.ma as ma
-from astropy import units as u
-from astropy.time import Time
+import logging
+from ctapipe.core import Map, Field
+from ctapipe.io.containers import(
+    R0CameraContainer,
+    R0Container,
+    DataContainer
+)
+from ctapipe.io.eventsource import EventSource
 
-from ctapipe.core import Container
-from .containers import R0Container
 
 logger = logging.getLogger(__name__)
 
-try:
-    import protozfitsreader
-except ImportError as err:
-    logger.fatal(
-        "the `protozfitsreader` python module is required to access MC data: {}"
-        .format(err))
-    raise err
-
-__all__ = [
-    'zfits_event_source',
-]
-
-
-def zfits_event_source(url, max_events=None):
-    """A generator that streams data from an EventIO/HESSIO MC data file
-    (e.g. a standard CTA data file.)
-
-    Parameters
-    ----------
-    url : str
-        path to file to open
-    max_events : int, optional
-        maximum number of events to read
-
-
+__all__ = ['ZFitsEventSource']
+
+
+class ZFitsEventSource(EventSource):
+    def _generator(self):
+        from protozfitsreader import ZFile
+        for count, event in enumerate(ZFile(self.input_url)):
+            data = SST1M_DataContainer()
+            data.count = count
+            data.r0.event_id = event.event_number
+            data.r0.tels_with_data = [event.telescope_id, ]
+
+            for tel_id in data.r0.tels_with_data:
+                r0 = data.r0.tel[tel_id]
+                r0.fill_from_zfile_event(event)
+            yield data
+
+    def is_compatible(self, path):
+        from astropy.io import fits
+        try:
+            h = fits.open(path)[1].header
+        except:
+            # not even a fits file
+            return False
+
+        # is it really a proto-zfits-file?
+        return (
+            (h['XTENSION'] == 'BINTABLE') and
+            (h['EXTNAME'] == 'Events') and
+            (h['ZTABLE'] is True) and
+            (h['ORIGIN'] == 'CTA') and
+            (h['PBFHEAD'] == 'DataModel.CameraEvent')
+        )
+
+    @property
+    def is_stream(self):
+        return True
+
+
+class SST1M_R0CameraContainer(R0CameraContainer):
```

Comment:
2 points:

Firstly, inheriting from `R0CameraContainer` does not inherit the Fields of `R0CameraContainer`.

Secondly, I think we should change the approach here. If there is event information you wish to store alongside the information in the default `R0CameraContainer`, that is not used in typical ctapipe execution, but could be useful for "expert/debugging" scripts, then it should be placed in its own container.

I suggest creating a new "SST1M" Container that is in addition to the default containers.

Reply:
> 

It does not? Why does this work then?
```python:
In [1]: from ctapipe.io.containers import R0CameraContainer
In [2]: class Foo(R0CameraContainer):
   ...:     pass
In [3]: foo = Foo()
In [4]: foo.adc_samples = np.arange(10)   # works
In [4]: foo.a = 3                                       # should not work
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-a25649f06fba> in <module>()
----> 1 foo.a = 3
AttributeError: 'Foo' object has no attribute 'a'
 # and it does not work. 
# Foo is pretty much exactly what I would call a R0CameraContainer
```

Reply:
> I think we should change the approach here. 

Can you make a proposal? I do not see how I can do this without inheriting from Containers, without repeating a lot of code.

Reply:
> It does not? Why does this work then?
Perhaps I am mistaken. I just remember encountering problems before when inheriting instead of creating from fresh.

> Can you make a proposal?

```python
class SST1MCameraContainer(Container):
    """
    Storage of raw data from a single telescope
    """
    pixel_flags = Field(ndarray, 'numpy array containing pixel flags')
    digicam_baseline = Field(ndarray, 'Baseline computed by DigiCam')
    local_camera_clock = Field(float, "camera timestamp")
    gps_time = Field(float, "gps timestamp")
    camera_event_type = Field(int, "camera event type")
    array_event_type = Field(int, "array event type")
    trigger_input_traces = Field(ndarray, "trigger patch trace (n_patches)")
    trigger_output_patch7 = Field(
        ndarray,
        "trigger 7 patch cluster trace (n_clusters)")
    trigger_output_patch19 = Field(
        ndarray,
        "trigger 19 patch cluster trace (n_clusters)")

    def fill_from_zfile_event(self, event):
        self.pixel_flags = event.pixel_flags
        self.digicam_baseline = event.baseline
        self.local_camera_clock = event.local_time
        self.gps_time = event.central_event_gps_time
        self.camera_event_type = event.camera_event_type
        self.array_event_type = event.array_event_type
        self.trigger_input_traces = event.trigger_input_traces
        self.trigger_output_patch7 = event.trigger_output_patch7
        self.trigger_output_patch19 = event.trigger_output_patch19

class SST1MContainer(Container):
    tel = Field(
        Map(SST1MCameraContainer),
        "map of tel_id to SST1MCameraContainer")


class SST1M_DataContainer(DataContainer):
   sst1m  = Field(SST1MContainer(), "SST1M Specific Information")
```

And then fill the R0 container separately. This ensures that the user can always expect the same contents to be found in the R0 Container, and not be confused about extra camera-specific information included. I think this separation is a better approach.